src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -395,21 +444,34 @@ public SocketConfig(EndPoint endPoint, int timeout, PingOptions? options, bool i
             public readonly byte[] SendBuffer;
         }
 
-        private static unsafe byte[] CreateSendMessageBuffer(IcmpHeader header, byte[] payload)
+        private static unsafe byte[] CreateSendMessageBuffer(IpHeader ipHeader, IcmpHeader icmpHeader, byte[] payload)
         {
-            int headerSize = sizeof(IcmpHeader);
-            byte[] result = new byte[headerSize + payload.Length];
-            Marshal.Copy(new IntPtr(&header), result, 0, headerSize);
-            payload.CopyTo(result, headerSize);
-            ushort checksum = ComputeBufferChecksum(result);
+            int icmpHeaderSize = sizeof(IcmpHeader);
+            int offset = 0;
+            int packetSize = ipHeader.TotalLength != 0 ? ipHeader.TotalLength : icmpHeaderSize + payload.Length;
```

Comment:
```suggestion
            int packetSize = ipHeader.TotalLength != 0 ? ipHeader.TotalLength : checked(icmpHeaderSize + payload.Length);
```
?  That way if it does overflow at least a more useful exception will be thrown.  Not a big deal, though.